elks/arch/i86/drivers/char/console-direct-pc98.c

```diff
@@ -48,27 +48,34 @@
 
 #define MAXPARMS	28
 
+#define MAX_OUTPUTS 1
+
 struct console;
 typedef struct console Console;
 
 struct console {
+    unsigned char op;
```

Comment:
Byte-sized struct members should not be placed at the top when not next to another byte member, structures are packed and this cause all subsequent word members to be odd-addressed, which greatly slows down CPU access on 8088.

Move new byte members either to the end of the struct or find another unaligned byte member to stick it next to.

What is "op", it is not clear to me? Is this the effective /dev/ttyN number minus 1?

Reply:
op = output, it was a bit more obvious when it was still `Output *op` in the previous iteration. That is now used as the index to the arrays `Visible`, `glock` and `glock_wait`, as we currently have one of each for every output.

Reply:
The above comment applies to all console*.c drivers; I just happened to notice this in PC-98 code.

Reply:
Still not obvious what "output" is.

> That is now used as the index to the arrays Visible, glock and glock_wait, as we currently have one of each for every output.

I'm not sure making any of the above glock/wait etc variables into arrays is needed - there's only one graphical screen that can be active per machine at a time. Most of this is extra code, all for nothing. Think simplicity at first - ELKS isn't designed to handle everthing, we are very limited in code/data space. I advise leaving glock/wait etc alone for the time being. They can be modified later when somebody tries to run two Nano-X applications simultaneously (which won't work anyways).

Reply:
Output is the display that console is on, so if C->op == 0, then it's on display 0. My original thinking was that we'd want per-output glock/glock_wait in the case that, say, both screens were running a program that wants the terminal in raw mode, for instance, since that lock also used to guard that.

> I'm not sure making any of the above glock/wait etc variables into arrays is needed - there's only one graphical screen that can be active per machine at a time.

In `console-direct.c` it was done to allow for multiple displays, and I changed it in the other ones because they all share the same `console.c` logic. They aren't really arrays in those, though, since `MAX_OUTPUTS` is defined as `1` in them.

Reply:
> Byte-sized struct members should not be placed at the top when not next to another byte member, structures are packed and this cause all subsequent word members to be odd-addressed, which greatly slows down CPU access on 8088.

Very good info! Noted.

Reply:
> They aren't really arrays in those, though, since MAX_OUTPUTS is defined as 1 in them.

Well, that doesn't matter - they're declared as arrays when they likely don't need to be, which is confusing, and the "array" is indexed by by compiled code, thus generating more code, even when C->op is zero. My point was that those variables are truly singleton system globals and don't need to be indexed, in any of the drivers, if this code is attempting to lock out graphics access on multiple monitors; it isn't needed.

Reply:
Okay, I'll restore those to how they were. We can obviously revisit them later if needed, and admittedly this change was added 'just in case' to begin with.

Reply:
I can see the approach you've taken - that of systematically following through with each global or other variable, and naturally creating arrays where there was a singleton, etc - and you've done a good job of it, and it works! 

However, for our kernel, we must always save bytes, data structure extensions, code generated, etc., so one must at all times keep that in mind - IIRC we're within 2-3k bytes of running out of kernel code space! You just happen to have taken on a somewhat complex enhancement as only your second enhancement. I don't normally comment this heavily on PRs - I am impressed with what you've done!

Reply:
> I don't normally comment this heavily on PRs - I am impressed with what you've done!

I don't remember the last time I enjoyed working on a PR this much! I really appreciate the thorough and informative feedback, so keep it coming! And don't hesitate to point out even small/trivial stuff, since this is only my second PR for this project, and it did end up a bit larger than I initially scoped :sweat_smile: 
If I missed something, I want to know!

One example is that I changed some instances of
```C
Console *C = Con;
```
to
```C
Console *C = &Con[0];
```

Because to me, that communicates the intent more precisely. But if you or any other maintainer of the project thinks the former way is better in any way, then I'm more than happy to fix that up. I try to avoid style changes when it's not my code I'm working on, but there is a good chance I've missed some spots, and those should be fixed as well.

So yeah, I'm really enjoying this :sunglasses:

Reply:
Your recent changes look very nice, thank you. Now that lots of detail is moved out into crtc-6845.c, I am starting to see the whole much more clearly. It's looking very good, there are a few small items I will comment on separately.

I am now also seeing the curious problem of why there were (are) Width vs MaxCol, when one is always literally one less than the other. Same for Height and MaxRow. So I don't like it lol :)

Obviously, getting in to changing these kinds of things can be very prone to human error, and I've already spent too much time solving off-by-one issues in the console driver over the past 4 years. But I have an idea:

First, I suggest we use the old names in struct Console: `MaxRow` instead of max_row, Width instead of w, etc.
That should allow for as few visual changes as possible in the large diff listing. And I definitely look at that stuff.

THEN, start replacing Width and Height (in the far fewer routines that use that versus MaxCol and MaxRow it seems) with a *local* variable: `unsigned int Width = C->MaxRow + 1`. Then just use the local Width rather than C->Width in the routine(s).

This may sound like overkill, but I think it allows us to (almost) prove that the large number of small changes in this PR won't in fact break the console, which will cause (me) to have to jump in and debug it, which I don't want to do! Also, the compiler may optimize and generate better code anyways using this approach, since only one struct member has to be referenced.

What do you think of that idea? I'm looking at routines like ScrollDown for best candidates for this kind of thing. If OTOH you think that it might be easier to swap and replace MaxCol with a local `unsigned int MaxCol = C->Width - 1`,  in console.c, then that's ok too. It's possible that two local variables MaxCol and MaxRow at the top of AnsiCmd will save a lot of code being generated!

Reply:
I actually had that change at some point to eliminate MaxCol and MaxRow, but I didn't want to stuff in more unrelated changes than necessary 😄

Still need to do that renaming, forgot to include that with the latest patches

Reply:
> I actually had that change at some point to eliminate MaxCol and MaxRow,

You're getting ahead of me - nice :) I understand about the hesitancy but now things are coming together nicely, thank you!